### PR TITLE
Ansible for installing Galaxy

### PR DESCRIPTION
**Pull request by erasche**
Not really ready for review yet

### CODE REVIEW

File: topics/admin/tutorials/ansible-galaxy/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+>    ```yaml
+>    - src: galaxyproject.galaxy
+>    - src: galaxyproject.postgresql
+>    - src: galaxyproject.proftpd
+>    - src: https://github.com/usegalaxy-eu/ansible-repos
+>      name: galaxyproject.repos
+>    - src: geerlingguy.nginx
+>    - src: natefoo.postgresql_objects
+>    - src: https://github.com/usegalaxy-eu/ansible-role-supervisor
+>      name: geerlingguy.supervisor
```
Comment (natefoo) [inline topics/admin/tutorials/ansible-galaxy/tutorial.md]: `usegalaxy-eu.supervisor`?

Comment (erasche) [reply topics/admin/tutorials/ansible-galaxy/tutorial.md]: figured I'd move it eventually?

**Comment by natefoo** [comment]
A few things that we might want to consider from a "best practices" standpoint:

1. We're making `/srv/galaxy` the `galaxy` user's home directory, meaning that...
1. `/srv/galaxy` is owned by `galaxy`, so `galaxy` has full write access to everything under it, which is not ideal (code, configs, etc. are preferably owned by another user e.g. `root`), which...
1. in our example are owned by `galaxy`. Also,
1. we're creating `/data` world readable. But to do otherwise would mean the student's regular user wouldn't be able to read the contents of the datasets. Ideally it'd be `750` and we'd add the student's user to the `galaxy` group, but that's more `pre_tasks`...

I know this is sort of a rabbit hole but since we decided to "teach the best practices" I wonder if we should incorporate these things, even if they end up adding a bunch of complexity?

One idea: Add user/dir creation to the `galaxyproject.galaxy` role that is, as with everything else, controllable via an Ansible variable...

**Comment by erasche** [comment]
> rabbit hole

Just a bit yeah haha. There's some tradeoff to be made, I'm not sure what the right balance is, but I think it's somewhere in between? We need to simultaneously:

- make sure that people can complete this training on their own (which I believe entails making sure that each step is only a small / incremental improvement rather than starting with too much complexity)
- as well as addressing best practices.

I'd support this if start from "best practices" and finish with "best security practices." Since it's easy for them to throw it all away, it wouldn't be too difficult to have config files owned as root at the end? 

> But to do otherwise would mean the student's regular user wouldn't be able to read the contents of the datasets

That's fine, they have to `su` to galaxy to do anything with supervisor, definitely had wrong premissions on the /data directory, thanks for fixing that in your branch! :)

> note it's broken right now

Yeah tried it out and noticed that. Looking at the readme for the playbook that you wrote, we'd have to do:

```
    - role: galaxy # as root
      galaxy_manage_mutable_setup: no
      galaxy_manage_database: no
    - role: galaxy
      become: yes
      become_user: galaxy
      galaxy_manage_clone: no
      galaxy_manage_static_setup: no
```

Starts to get a bit scary in terms of what we'd have to explain. I'd be much more comfortable to start with "run the role once", and near the end we talk about securing it. It would fit with the planned section for adding SSL, we can cover best practices with some discussion, why you should own code with a different user, here is how you can accomplish that easily.

But to introduce this from the start, I fear that it's a lot of mental complexity to handle when they've just learnt ansible that very morning?

File: topics/admin/tutorials/ansible-galaxy/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+>
+{: .agenda}
+
+
+# Playbook Overview
+
+
+## Configuration
+
+The official playbook is extremely configurable, everything that you want to change is exposed as a variable, and then tasks will change behaviour based on that. The [playbook documentation](https://github.com/galaxyproject/ansible-galaxy#role-variables) is the most up-to-date source of documentation for the variables. You should take a minute a read over the variables listed there.
```
[the remark this answers is not in the document]
Comment (nsoranzo) [reply topics/admin/tutorials/ansible-galaxy/tutorial.md]: > take a minute a read over the variables

This doesn't sound right.

File: topics/admin/tutorials/ansible-galaxy/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,849 @@
...
+    link: "https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html"
+  -
+    title: Comfort with Ansible roles and tasks
+    type: "none"
+---
+
+# Overview
+{:.no_toc}
+
+Now that you have some familiarity with Ansible and are comfortable running existing playbooks that you've written, we'll move on to installing Galaxy with a playbook.
```
Comment (nsoranzo) [inline topics/admin/tutorials/ansible-galaxy/tutorial.md]: This intro is a bit too tailored for the Admin course, can you rephrase it so it applies also for people reading this on their own (and which may have learnt Ansible somewhere else)?

File: topics/admin/tutorials/ansible-galaxy/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,849 @@
...
+
+
+# Playbook Overview
+
+
+## Configuration
+
+The official playbook is extremely configurable, everything that you want to change is exposed as a variable, and then tasks will change behaviour based on that. The [playbook documentation](https://github.com/galaxyproject/ansible-galaxy#role-variables) is the most up-to-date source of documentation for the variables. You should take a minute a read over the variables listed there.
+
+The important variables for our purposes today will be:
```
Comment (nsoranzo) [inline topics/admin/tutorials/ansible-galaxy/tutorial.md]:
```suggestion
The important variables for this tutorial are:
```

File: topics/admin/tutorials/ansible-galaxy/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,849 @@
...
+
+- `galaxy_server_dir`
+- `galaxy_commit_id`
+- `galaxy_config`
+
+These are largely self explanatory, a server directory, which commit should be installed, and the Galaxy configuration. We will not Galaxy configuration variables in detail as they are covered sufficiently in the `galaxy.yml` file or the [online documentation](https://docs.galaxyproject.org/en/master/admin/config.html#configuration-options).
```
Comment (nsoranzo) [inline topics/admin/tutorials/ansible-galaxy/tutorial.md]:
```suggestion
These are largely self explanatory: a server directory, which commit should be installed, and the Galaxy configuration. We will not explain Galaxy configuration variables in detail as they are covered sufficiently in the `galaxy.yml` sample file or the [online documentation](https://docs.galaxyproject.org/en/master/admin/config.html#configuration-options).
```

File: topics/admin/tutorials/ansible-galaxy/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,849 @@
...
+
+1. It starts by checking if the Galaxy directory exists, this is done so the tasks will attempt to check the current commit, will not fail in case the folder does not exist.
+2. The current commit of the repository on disk is retrieved, in order to decide whether or not Galaxy needs to be updated. This is then reported in a debugging message to the user.
+3. Galaxy is updated to the correct commit. If the `galaxy_server_dir` is not yet populated, e.g. on first run, Galaxy will be cloned to the specified location. Otherwise the existing repo will be updated to that commit. The `galaxy_force_checkout` variable allows you to control whether or not any locally made changes will be discarded. This will not affect mutable configuration files as those are not tracked in git.
+4. The virtualenv is set up:
+	1. An empty virtualenv is created.
+	2. Pip is updated within the virtualenv.
+5. Any `.pyc` files are removed, as this can occasionally result in python loading the cached code, instead of any updates that were made by changing commit. For safety, all of these are removed.
```
Comment (nsoranzo) [inline topics/admin/tutorials/ansible-galaxy/tutorial.md]:
```suggestion
5. Any `.pyc` files are removed, as this can occasionally result in Python loading the cached code, even if the corresponding `.py` file is no more present at the checked-out commit. For safety, all of these are removed.
```

File: topics/admin/tutorials/ansible-galaxy/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,849 @@
...
+
+With that Galaxy is cloned to disk and is ready to be configured by the next task
```
Comment (nsoranzo) [inline topics/admin/tutorials/ansible-galaxy/tutorial.md]:
```suggestion
With that Galaxy is cloned to disk and is ready to be configured by the next task.
```

**Comment by natefoo** [comment]
@erasche no disagreements from me with anything you said.

I've always hated running the role twice (3 times in some cases!). I be we could instead put task-specific (or include-specific) `become` and `become_user` on certain tasks that are controllable via variables. I'm going to spend a few minutes on the galaxy role to see if this would solve the whole thing.

File: topics/admin/tutorials/ansible-galaxy/tutorial.md (hunk 3 above)
Comment (erasche) [reply topics/admin/tutorials/ansible-galaxy/tutorial.md]: This tutorial assumes the knowledge contained in the requirements, namely the existing ansible tutorial in the training material.

**Comment by erasche** [comment]
> no disagreements from me with anything you said.

toll :)

> I'm going to spend a few minutes on the galaxy role to see if this would solve the whole thing.

klinkt gut!!

File: topics/admin/tutorials/ansible-galaxy/tutorial.md (hunk 3 above)
Comment (nsoranzo) [reply topics/admin/tutorials/ansible-galaxy/tutorial.md]: I've seen that, what I don't like of the present formulation is the "Now that you have..." and "we'll move on" parts, which seem to assume you have just finished the Ansible tutorial, while the reader may have learnt it somewhere else or a long time ago.

What about the following:

```suggestion
This tutorial assumes you have some familiarity with Ansible and are comfortable with writing and running playbooks. Here we'll see how to install a Galaxy server using an Ansible playbook.
```

Comment (erasche) [reply topics/admin/tutorials/ansible-galaxy/tutorial.md]: Thanks for the suggestion, yeah, that fits for me :) Definitely needed to rework that formulation, don't want to put too much pressure on the reader, just that "there is an assumption/belief that you're comfortable with this already."

**Comment by erasche** [comment]
@nsoranzo will make a final pass for language (not just grammar/flow, but also that it doesn't make knowledge assumptions, or make the reader ever feel bad) just before out of WIP, would appreciate your review after that if you have time.

File: topics/admin/tutorials/ansible-galaxy/galaxyservers.txt [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    # Our additions
+    mule:
+      - lib/galaxy/main.py
+      - lib/galaxy/main.py
+    farm: job-handlers:1,2
+
+# Supervisord
+supervisor_socket_user: 'galaxy'
+supervisor_socket_chown: 'galaxy'
+
```
Comment (Slugger70) [inline topics/admin/tutorials/ansible-galaxy/galaxyservers.txt]: `galaxy_restart_handler_name: restart galaxy`

**Comment by shiltemann** [comment]
thanks all, this is great, I've wanted this tutorial for myself for a long time! :D
